### PR TITLE
Rd 5590 handle node tracer scrubbing issues

### PR DESCRIPTION
**Pull request by OrrLumigo**

### CODE REVIEW

**Review by efimk-lu — APPROVED** [review]

**Review by saartochner — APPROVED** [review]
Can you add test?

File: src/reporter.test.ts [hunk 1]
```diff
@@ -121,11 +123,352 @@ describe('reporter', () => {
...
+        },
+        { dummyEnd },
+      ];
+      const expectedResultSize = getJSONBase64Size(spans);
+
+      const actual = JSON.parse(reporter.forgeAndScrubRequestBody(spans, expectedResultSize));
+      expect(actual).toEqual(expected);
+    });
+
+    test('getHttpInfo contain json header but not json body', () => {
```
Comment (nirLumigo) [inline src/reporter.test.ts:421]: You don't run the function `getHttpInfo` in the test

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Nice! :)

File: src/reporter.ts [hunk 2]
```diff
@@ -45,38 +51,101 @@ export const shouldTrim = (spans, maxSendBytes: number): boolean => {
...
+    if (isJsonContent(payload, headers) && isContainingSecrets(payload)) {
+      if (!(payload.length < sizeLimit)) {
+        payload = untruncateJson(payload);
+      }
+      return payloadStringify(JSON.parse(payload), sizeLimit);
+    } else {
+      return payloadStringify(payload, sizeLimit);
+    }
+  } catch (e) {
+    return payloadStringify(payload, sizeLimit);
```
[the remark this answers is not in the document]
Comment (OrrLumigo) [reply src/reporter.ts:69]: this is what was before also

[the remark this answers is not in the document]
Comment (OrrLumigo) [reply src/reporter.ts]: Not exactly

File: src/reporter.test.ts [hunk 3]
```diff
@@ -467,6 +468,80 @@ describe('reporter', () => {
       const actual = JSON.parse(reporter.forgeAndScrubRequestBody(spans, expectedResultSize));
       expect(actual).toEqual(expected);
     });
+
+    test('getHttpSpan - response with error should double payload size', () => {
```
Comment (nirLumigo) [inline src/reporter.test.ts:472]: You don't test here the function `getHttpSpan`

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Nice :)

File: src/globals.ts [hunk 4]
```diff
@@ -10,10 +11,13 @@ export const DEFAULT_TRACER_TIMEOUT = 500;
 
 export const SpansContainer = (() => {
   let spansToSend = {};
-
+  let totalSize = 0;
   const addSpan = (span) => {
-    spansToSend[span.id] = span;
-    logger.debug('Span created', span);
+    if (spanHasErrors(span) || getMaxRequestSize() > totalSize) {
```
[the remark this answers is not in the document]
Comment (OrrLumigo) [reply src/globals.ts:16]: it has nothing to do with size
we want to stop adding spans after a certain total size of spans but always collect errors

File: src/reporter.ts [hunk 5]
```diff
@@ -105,47 +105,27 @@ function scrubSpans(resultSpans: any[]) {
...
-  if (spans.length - resultSpans.length > 0) {
+  scrubSpans(spans);
+  if (originalSize - spans.length > 0) {
     logger.debug(`Trimmed spans due to size`);
   }
   logger.debug(
-    `Filtered [${spans.length - resultSpans.length}] spans out, Took: [${
-      new Date().getTime() - start
-    }ms]`
+    `Filtered [${spans.length - spans.length}] spans out, Took: [${new Date().getTime() - start}ms]`
```
[the remark this answers is not in the document]
Comment (OrrLumigo) [reply src/reporter.ts]: ?

**Review by nirLumigo — APPROVED** [review]

File: src/hooks/http.ts [hunk 6]
```diff
@@ -264,32 +264,39 @@ export class Http {
...
       if (args[0] === 'end') {
+        let maxSizeNoErrors = getEventEntitySize();
         const responseData = {
           statusCode,
           receivedTime,
-          body,
+          body:
+            statusCode < 400 && body.length > maxSizeNoErrors
+              ? body.substr(0, maxSizeNoErrors)
+              : body,
```
Comment (doriaviram) [inline src/hooks/http.ts:290]: Are we taking the all body in this case? even if it's 1GB?

File: src/reporter.ts [hunk 7]
```diff
@@ -36,7 +42,7 @@ export const sendSpans = async (spans): Promise<void> => {
   const roundTripEnd = Date.now();
   const rtt = roundTripEnd - roundTripStart;
 
-  logSpans(rtt, spans);
+  logSpans(rtt, spans.slice(Math.min(25, spans.length)));
```
Comment (doriaviram) [inline src/reporter.ts:45]: Export this `25` to a variable

Comment (doriaviram) [reply src/reporter.ts:45]: Why are we doing that ? we don't want to limit the spans to `25`

**Review by doriaviram — CHANGES_REQUESTED** [review]
Lets talk about this change

File: src/hooks/http.ts (hunk 6 above)
Comment (OrrLumigo) [reply src/hooks/http.ts:290]: no there is a max of 2048KB with no error and 4096KB with errors

Comment (OrrLumigo) [reply src/hooks/http.ts:290]: if (chunk.length > allowedLengthToAdd) { 

makes sure of that

File: src/globals.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (OrrLumigo) [reply src/globals.ts:17]: errors always go through

File: src/reporter.ts (hunk 7 above)
Comment (doriaviram) [inline src/reporter.ts:45]: Log them all

File: src/reporter.ts [hunk 8]
```diff
@@ -45,38 +51,84 @@ export const shouldTrim = (spans, maxSendBytes: number): boolean => {
...
+  } catch (e) {
+    return payloadStringify(payload, sizeLimit);
   }
-
-  resultSpans.push(functionEndSpan);
-
-  if (spans.length - resultSpans.length > 0) {
+}
+
+function scrubSpans(resultSpans: any[]) {
```
Comment (doriaviram) [inline src/reporter.ts:71]: Im not sure this is the right place for this function, this is should be only the reporter, scrubbing is part of the collection

File: src/reporter.ts [hunk 9]
```diff
@@ -45,38 +51,84 @@ export const shouldTrim = (spans, maxSendBytes: number): boolean => {
...
+        span.info.httpInfo.request.headers = payloadStringify(request.headers, sizeLimit);
+        if (response.headers)
+          span.info.httpInfo.response.headers = payloadStringify(response.headers, sizeLimit);
+      }
+    }
+  });
+}
+
+export const forgeAndScrubRequestBody = (spans, maxSendBytes): string | undefined => {
+  const start = new Date().getTime();
```
Comment (doriaviram) [inline src/reporter.ts:112]: This function is not pure, can we make it pure ?

File: src/reporter.ts [hunk 10]
```diff
@@ -45,38 +51,84 @@ export const shouldTrim = (spans, maxSendBytes: number): boolean => {
...
+    );
+
+    const functionEndSpan = spans.pop();
+    spans.sort((a, b) => (spanHasErrors(a) ? -1 : spanHasErrors(b) ? 1 : 0));
+    let totalSize = getJSONBase64Size(functionEndSpan) + getJSONBase64Size(spans);
+    while (totalSize > maxSendBytes && spans.length > 0)
+      totalSize -= getJSONBase64Size(spans.pop());
+    spans.push(functionEndSpan);
+  }
+  scrubSpans(spans);
```
Comment (doriaviram) [inline src/reporter.ts:126]: Are we not scrubbing in the collection anymore? only while reporting

File: src/reporter.ts [hunk 11]
```diff
@@ -45,38 +51,84 @@ export const shouldTrim = (spans, maxSendBytes: number): boolean => {
...
+          );
+        }
+        span.info.httpInfo.request.headers = payloadStringify(request.headers, sizeLimit);
+        if (response.headers)
+          span.info.httpInfo.response.headers = payloadStringify(response.headers, sizeLimit);
+      }
+    }
+  });
+}
+
```
Comment (OrrLumigo) [inline src/reporter.ts:110]: add note that this alters spans to save on memory

File: src/hooks/http.ts [hunk 12]
```diff
@@ -264,32 +265,41 @@ export class Http {
     response
   ) {
     let body = '';
-    const payloadSize = getEventEntitySize();
+    let maxPayloadSize = getEventEntitySize(true);
```
Comment (doriaviram) [inline src/hooks/http.ts:268]: its still const

**Review by doriaviram — APPROVED** [review]

**Pull request merged**